File: 131-palindrome-partitioning/131-palindrome-partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        global total, res
        total = []
        res = []
        def is_palindrome(string):
            left, right = 0, len(string) - 1
            while(left < right):
                if string[left] != string[right]:
                    return False
                else:
                    left += 1
                    right = right - 1
            return True
        length = len(s)
        def dfs(s):
            if len(s) == 0:
                total.append(res[:])
                return
            for i in range(len(s)):
                substring = s[:(i+1)]
                if is_palindrome(substring):
                    res.append(substring)
                    dfs(s[(i+1):])
                    res.pop()
        dfs(s)
        return total
            

File: 131-palindrome-partitioning/131-palindrome-partitioning.java
class Solution {
    public List<List<String>> partition(String s) {
        //@input : string s
        //@output: all possible palindrome partitioning of s
        //@edge case : no
        //@breaking down problem:
        // 1. backtracking
        // 2. if come to the end, then terminate
        // 3. in each step, select the palindrome one
        List<List<String>> res = new LinkedList<>();
        List<String> temp = new LinkedList<>();
        backtracking(s, 0, res, temp);
        return res;
        
        
        
    }
    public boolean isPalindrome(String s){
        int n = s.length();
        int i = 0;
        while(i < n / 2){
            if(s.charAt(i) != s.charAt(n - 1 - i))return false;
            i++;
        }
        return true;
    }
    
    public void backtracking(String s, int curPointer, List<List<String>> res, List<String> temp){
        //termination
        
        if(curPointer == s.length()){
            res.add(new LinkedList<>(temp));
            return;
        }
        for(int i = curPointer; i < s.length(); i++){
            //select
            String curPartition = s.substring(curPointer, i + 1);
            if(isPalindrome(curPartition)){
                temp.add(curPartition);
                backtracking(s, i + 1, res, temp);
                temp.remove(temp.size() - 1);
            }
        }
    }
    
    
}